vaadin-spring/src/main/java/com/vaadin/flow/spring/scopes/VaadinRouteScope.java

```diff
@@ -0,0 +1,380 @@
+        public void onComponentEvent(DetachEvent event) {
+            assert getVaadinSession().hasLock();
+            uiDetachRegistration.remove();
+            if (resetUI()) {
+                uiDetachRegistration = currentUI.addDetachListener(this);
+            } else {
+                destroy();
+            }
+        }
+
+        @Override
+        protected Object doGet(String name, ObjectFactory<?> objectFactory) {
+            RouteScopeOwner owner = getContext().findAnnotationOnBean(name,
+                    RouteScopeOwner.class);
+            if (owner == null) {
+                return objectFactory.getObject();
+            }
+
+            if (!getNavigationListener().hasNavigationOwner(owner)) {
+                throw new IllegalStateException(String.format(
+                        "Route owner '%s' instance is not available in the "
+                                + "active navigaiton components chain: the scope defined by the bean '%s' doesn't exist.",
+                        owner.value(), name));
+            }
+            return super.doGet(name, objectFactory);
+        }
+
+        @Override
+        protected void storeBean(String name, Object bean) {
+            super.storeBean(name, bean);
+            RouteScopeOwner owner = getContext().findAnnotationOnBean(name,
+                    RouteScopeOwner.class);
+            getNavigationListener().storeOwner(name, owner.value());
+        }
+
+        private boolean resetUI() {
+            for (UI ui : getVaadinSession().getUIs()) {
+                String windowName = getWindowName(ui);
+                if (ui != currentUI
+                        && windowName.equals(getWindowName(currentUI))) {
+                    currentUI = ui;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private ApplicationContext getContext() {
+            VaadinService service = currentUI.getSession().getService();
+            VaadinContext context = service.getContext();
+            ServletContext servletContext = ((VaadinServletContext) context)
+                    .getContext();
+            return WebApplicationContextUtils
+                    .getWebApplicationContext(servletContext);
+        }
+
+        private NavigationListener getNavigationListener() {
+            return ComponentUtil.getData(currentUI, NavigationListener.class);
+        }
+
+    }
+
+    @Override
+    public void postProcessBeanFactory(
+            ConfigurableListableBeanFactory beanFactory) {
+        beanFactory.registerScope(VAADIN_ROUTE_SCOPE_NAME, this);
+    }
+
+    @Override
+    public String getConversationId() {
+        return getVaadinSession().getSession().getId() + "-route-scope";
+    }
+
+    @Override
+    public void uiInit(UIInitEvent event) {
+        NavigationListener listener = new NavigationListener(event.getUI());
+        ComponentUtil.setData(event.getUI(), NavigationListener.class,
```

Comment:
Should the data be cleared at some point to reduce the risk that detached UIs would remain referenced from the VaadinRouteScope or RouteBeanStore ? I don't think it is referenced (and thus leaked) in this PR, but there is a risk that some future change could start leaking memory (at least in theory).

Reply:
I don't understand this comment.

It looks like it's attached to `ComponentUtil.setData(event.getUI(), NavigationListener.class,` line.
This line stores a navigation listener which is 1:1 to `UI` into the `UI`.
This is a cyclic reference which will be properly GCed.

I don't see how are `VaadinRouteScope` and `RouteBeanStore` applicable here.

Reply:
`VaadinRouteScope` is a singleton bean and it doesn't store anything inside itself.
`RouteBeanStore` is 1:1 to UI and also will be properly GCed.

The only valid point is `RouteStoreWrapper` may keep extra data per `UI` which can be detached.
But: 
- it's data without the `UI` itself
- it's stored per session and will be removed on session remove.
- it's not a place to care about `RouteStoreWrapper` .

But I've fixed this case.